Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -30,275 +30,1335 @@ pub type SizeT = usize;
+                let borrowed: Option<&Rc<Vec<T>>> = clone_boxed_borrowed.as_ref();
+                if let Some(cache) = borrowed.as_ref() {
+                    return Rc::clone(cache);
+                }
+                // Let's create an array of size length and fill it up recursively
+                // We don't materialize nested arrays because most of the time they are forgotten
+                let mut array: Vec<T> = Vec::with_capacity(*length);
+                Sequence::<T>::append_recursive(&mut array, self);
+                let result = Rc::new(array);
+                let mut cache = boxed.borrow_mut();
+                let mutable_left: *mut Sequence<T> = left.get();
+                let mutable_right: *mut Sequence<T> = right.get();
+                unsafe { *mutable_left = seq!() };
+                unsafe { *mutable_right = seq!() };
+                *cache = Some(result.clone());
+                result
+            }
         }
     }
 
-    fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
+    pub fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
         match this {
             Sequence::ArraySequence { values, .. } =>
               // The length of the elements
               for value in values.iter() {
                 array.push(value.clone());
               },
-            Sequence::ConcatSequence { left, right, .. } =>
+            Sequence::ConcatSequence { boxed, left, right, .. } =>
               // Let's create an array of size length and fill it up recursively
               {
-                Sequence::<T>::append_recursive(array, left);
-                Sequence::<T>::append_recursive(array, right);
+                let clone_boxed = boxed.as_ref().clone();
+                let clone_boxed_borrowed = clone_boxed.borrow();
+                let borrowed: Option<&Rc<Vec<T>>> = clone_boxed_borrowed.as_ref();
+                if let Some(values) = borrowed.as_ref() {
+                    for value in values.iter() {
+                        array.push(value.clone());
+                    }
+                    return;
+                }
+                Sequence::<T>::append_recursive(array, unsafe { &mut *left.get() });
+                Sequence::<T>::append_recursive(array, unsafe { &mut *right.get() });
               }
-            Sequence::LazySequence { boxed, .. } =>
-            Sequence::<T>::append_recursive(array, boxed.borrow().as_ref()),
         }
     }
-
     /// Returns the cardinality of this [`Sequence<T>`].
     // The cardinality returns the length of the sequence
-    fn cardinality(&self) -> SizeT {
+    pub fn cardinality_usize(&self) -> SizeT {
         match self {
             Sequence::ArraySequence { values, .. } =>
               // The length of the elements
               values.len(),
             Sequence::ConcatSequence { length, .. } =>
               *length,
-            Sequence::LazySequence { length, .. } =>
```

Comment:
Why were you able to remove this?

Reply:
I moved the lazy refcell into Concat itself, so there is no need for a third LazySequence.